File: 2018v/identicon/Lisachenko/identicon.py
import random
import sys

N = 5
M = 20

def rect(x, y, width, height, col_r, col_g, col_b):
    print('<rect x="' + str(x) + '" y="' + str(y) + '" width="' + str(width) + '" height="' + str(height) + '" style="fill:rgb('+str(col_r)+','+str(col_g)+','+str(col_b)+')" />')

def circle(cx, cy, r, col_r, col_g, col_b):
    print('<circle cx="' + str(cx) + '" cy="' + str(cy) + '" r="' + str(r) + '" style="fill:rgb('+str(col_r)+','+str(col_g)+','+str(col_b)+')" />')

def identicon(s, f):
    im = [[0 for i in range(N)] for j in range(N)]
    random.seed(s)
    
    for i in range(N):
        for j in range((N + 1) // 2):
            im[i][j] = int(random.random() > 0.5)
            im[i][-j - 1] = im[i][j]
    
    print('<p>' + s + '</p>')
    print('<svg width="' + str(M*N) + '" height="' + str(M*N) + '">')
    
    r=255
    g=255
    b=255
    while r + b + g > 500:
        r = random.randint(0, 255)
        g = random.randint(0, 255)
        b = random.randint(0, 255)
    
    if f == 'r':
        for i in range(N):
            for j in range(N):
                if im[i][j]:
                    rect(M*j, M*i, M, M, r, g, b)
    
    elif f == 'c' or f == 'c++':
        for i in range(N):
            for j in range(N):
                if im[i][j]:
                    circle(M*j + M//2, M*i + M//2, M//2, r, g, b)
                    if i != 0 and im[i-1][j]:
                        rect(M*j, M*i - M//2, M, M, r, g, b)
                    if j != 0 and im[i][j-1]:
                        rect(M*j - M//2, M*i, M, M, r, g, b)
                if not im[i][j] and f == 'c++':
                    if i != 0 and j != 0 and im[i-1][j] and im[i-1][j-1] and im[i][j-1]:
                        rect(M*j - M//2, M*i - M//2, M, M, r, g, b)
                    if i != N-1 and j != 0 and im[i+1][j] and im[i+1][j-1] and im[i][j-1]:
                        rect(M*j - M//2, M*i + M//2, M, M, r, g, b)                    
                    if i != 0 and j != N-1 and im[i-1][j] and im[i-1][j+1] and im[i][j+1]:
                        rect(M*j + M//2, M*i - M//2, M, M, r, g, b)
                    if i != N-1 and j != N-1 and im[i+1][j] and im[i+1][j+1] and im[i][j+1]:
                        rect(M*j + M//2, M*i + M//2, M, M, r, g, b)
                    circle(M*j + M//2, M*i + M//2, M//2, 255, 255, 255)


print('<html>\n<body>\n')
identicon(sys.argv[1], 'c++')
print('\n</html>\n</body>\n')

File: 2018a/identicon/reznik/Identicon.java
package olympiad;

import java.util.BitSet;
import java.util.Random;
import java.util.Scanner;

public class Identicon {

	public static final int SIZE = 32;
	public static final int COLOR = 3;
	public static final int LOCATION = 32;

	public static int[] chooseColor(int color) {
		int[] a = new int[3];
		if (color == 0) {// Black
			a[0] = 0;
			a[1] = 0;
			a[2] = 0;
		}
		if (color == 1) {// Red
			a[0] = 255;
			a[1] = 0;
			a[2] = 0;
		}
		if (color == 2) {// Orange
			a[0] = 255;
			a[1] = 140;
			a[2] = 0;
		}
		if (color == 7) {// Yellow
			a[0] = 255;
			a[1] = 255;
			a[2] = 0;
		}
		if (color == 4) {// Green
			a[0] = 0;
			a[1] = 255;
			a[2] = 0;
		}
		if (color == 5) {// Cyan
			a[0] = 0;
			a[1] = 255;
			a[2] = 255;
		}
		if (color == 6) {// Blue
			a[0] = 0;
			a[1] = 0;
			a[2] = 255;
		}
		if (color == 3) {// Purple
			a[0] = 255;
			a[1] = 0;
			a[2] = 255;
		}
		return a;
	}

	public static void drawIdenticon(BitSet field, int[] rgbColor) {
		int side = 7;
		int halfSide = (int)Math.ceil((double)side/2.0);
		int color = COLOR;
		String colour = rgbColor[0] + ","+rgbColor[1]+","+rgbColor[2];
		for (int i = 0; i < side; i++) {
			for (int j = 0; j < halfSide; j++) {
				if(field.get(color + halfSide*i + j)) {
					System.out.println("<rect x = \"" + 20*j + "\" y = \"" + 20*i + "\" width=\"20\" height=\"20\" style=\"fill:rgb("+colour+")\" />");
				}
			}
		}
		for (int i = 0; i < side; i++) {
			for (int j = 0; j < halfSide-1; j++) {
				if(field.get(color + halfSide*i + j)) {
					System.out.println("<rect x = \"" + (120 - 20*j) + "\" y = \"" + 20*i + "\" width=\"20\" height=\"20\" style=\"fill:rgb("+colour+")\" />");
				}	
			}
		}
	}


	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		Random rnd = new Random(in.nextLine().hashCode());
		long seed = rnd.nextLong();
		BitSet field = BitSet.valueOf(new long[]{seed});
		System.out.println(field.toString());
		int color = 0;
		if (field.get(0)) {
			color += Math.pow(2, COLOR - 1);
		}
		if (field.get(1)) {
			color += Math.pow(2, COLOR - 2);
		}
		if (field.get(2)) {
			color++;
		}
		int[] rgbColor = chooseColor(color);
		System.out.println("<html><body><svg width=\"140\" height=\"140\">");
		drawIdenticon(field, rgbColor);
		System.out.println("</svg></body></html>");
	}

}

